### PR TITLE
refactor(tracing): patch all integrations on import

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

This change ensures that all integrations patch the relevant modules only when they are imported. This is to reduce the potential side effects that importing modules might cause in the tracer. Most notably, importing modules too early might conflict with frameworks, such as gevent, that do their own monkey patching.

### Current Workflow for ddtrace-run
The current workflow in using ddtrace-run to run tracing involves a sitecustomize script and our `_monkey.py` module to auto-patch instrumented modules. This workflow is briefly summarized below:

1. ddtrace-run command
Our recommended usage of tracing a user's application is to run the executable application along with `ddtrace-run` as a prefix, like so:

```shell
ddtrace-run <user_application_command>
```
- This runs the `ddtrace_run.py` script, which adds the bootstrap directory containing our `sitecustomize.py` script to the interpreter's PYTHONPATH.
- This script then runs the executable `<user_application_command>` as normal, but due to Python being Python, the interpreter imports the bootstrap `sitecustomize.py` script during initialization before the executable is run.

2. Bootstrap Sitecustomize.py script
- The `sitecustomize.py` script initializes the tracer, profiler, dynamic instrumentation, runtime metrics, and configures the tracer with any environment variables (priority sampling, debug mode)
- The script then checks for any user-specified modules to patch via `DD_PATCH_MODULES=...`, and adds to our `EXTRA_PATCHED_MODULES` (currently containing web framework integrations).
- The script then makes a call to `patch_all()` in `_monkey.py` with the user-specified and web patched modules, and sets tracer-level tags via envrionment variables including `DD_ENV`, `DD_TRACE_GLOBAL_TAGS`
- Finally, the script then checks for and imports any other sitecustomize scripts that would have been used normally without `ddtrace-run`.

3. _monkey.py `patch_all()`
- The `_monkey.py` file has a default list of modules to patch, which gets overridden by any user specified environment variables `DD_TRACE_*_ENABLED` and `DD_PATCH_MODULES`
- `patch_all()` patches modules on import by registering the integration patch functions to after the module is imported, or runs the integration patch function right away.

### Proposed Workflow 

1. ddtrace-run command: Same as original workflow.
2. Bootstrap Sitecustomize.py script: Same as original workflow.
3. _monkey.py `patch_all()`:
`patch_all()` isn't impacted as much as `patch()`. `patch()` no longer imports and runs the integration patch functions immediately. Now, it registers the importing and running of the integration patch functions as a hook specifically on import of the module itself.

### Risk

The biggest risk inherent in this change is that it breaks some integrations in subtle ways. Hopefully our existing test suite covers the majority of such cases.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.

### CODE REVIEW

File: .github/workflows/test_frameworks.yml [hunk 1]
```diff
@@ -53,11 +53,22 @@ jobs:
         run: PYTHONPATH=../ddtrace/tests/debugging/exploration/ ddtrace-run pytest test --continue-on-collection-errors -v -k 'not test_simple'
 
   django-testsuite-3_1:
+    strategy:
```
Comment (Kyle-Verhoog) [inline .github/workflows/test_frameworks.yml:56]: This seems like a good change but is outside the scope of this PR

**Review by Kyle-Verhoog — COMMENTED** [review]
High-level the approach looks good.

We should have tests that verify `ddtrace-run` usage for each integration as we are now patching on import so that we have confidence in this change.

**Comment by P403n1x87** [comment]
> High-level the approach looks good.

🙌 

> We should have tests that verify `ddtrace-run` usage for each integration as we are now patching on import so that we have confidence in this change.

As we discussed off-line, the `PatchTestCase.Base` might be a good place to add a new test case that applies to all integrations. I haven't looked into what this might look like yet though, so I'm very much open to suggestions! 🙂

**Review by Yun-Kim — COMMENTED** [review]
I have some silly questions just looking for some clarification. Also, I can add a brief outline on our current ddtrace-run patching strategy but if you could also add some details on this new proposed patching strategy in the description that would be great! 😄

File: ddtrace/ext/sql.py [hunk 2]
```diff
@@ -20,15 +20,11 @@ def normalize_vendor(vendor):
         return vendor
 
 
-try:
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/ext/sql.py:23]: Here we are in a utility library and therefore we shouldn't import from more specific code. With patch-on-import, this means that importing from here actually triggers the import of the `psycopg` integration, causing a circular dependency.

File: ddtrace/_monkey.py [hunk 3]
```diff
@@ -210,98 +218,35 @@ def patch(raise_errors=True, patch_modules_prefix=DEFAULT_MODULES_PREFIX, **patc
...
-
-            # manually add module to patched modules
-            with _LOCK:
-                _PATCHED_MODULES.add(module)
-        else:
-            _patch_module(module, patch_modules_prefix=patch_modules_prefix, raise_errors=raise_errors)
+    contribs = [c for c, should_patch in patch_modules.items() if should_patch]
+    for contrib in contribs:
+        # Check if we have the requested contrib.
+        if not os.path.isfile(os.path.join(os.path.dirname(__file__), "contrib", contrib, "__init__.py")):
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/_monkey.py:224]: 🤔 I think that was a different error for when we couldn't find the `patch` function. Here we raise `ModuleNotFoundException` instead because we failed to find the contrib module.

File: ddtrace/contrib/cassandra/session.py [hunk 4]
```diff
@@ -3,7 +3,11 @@
 """
 import sys
 
-import cassandra.cluster
+
+try:
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/cassandra/session.py:7]: In Python < 3.7, `import cassandra.cluster` and similar imports give an attribute error saying that `cassandra` has no attribute `cluster`. If that happens, we try importing differently, i.e. with the `from ... import` form.

File: ddtrace/contrib/django/patch.py [hunk 5]
```diff
@@ -59,7 +59,22 @@
 )
 
 
+_NotSet = object()
+psycopg_cursor_cls = Psycopg2TracedCursor = _NotSet
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/django/patch.py:63]: This was another example of cross-importing from contribs, which could (and in this case it did) lead to circular imports. By the time `patch_conn` is called, `psycopg2` should be fully loaded and we can then try importing.

File: ddtrace/_monkey.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/_monkey.py:224]: Okay, then maybe we can remove the `IntegrationNotAvailableException` class since that is no longer being used.

File: ddtrace/contrib/cassandra/session.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/contrib/cassandra/session.py:7]: Does this mean that this import has been failing for Python <=3.6 for all cases? If so, should we separate this change out into a separate PR?

**Review by Yun-Kim — COMMENTED** [review]
Another silly question since I know we've moved futures as a contrib dependency in _monkey.py: why are we removing this compat module here and removed calls to patch futures in the tornado integration?

File: ddtrace/contrib/tornado/compat.py [hunk 6]
```diff
@@ -1,16 +0,0 @@
-try:
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/tornado/compat.py:1]: Because of the way that futures is checked for. Importing from another contrib should be considered an anti-pattern. Contribs should only be allowed to import from generic utilities, but not within each other.

File: ddtrace/_monkey.py [hunk 7]
```diff
@@ -84,24 +82,22 @@
     "dogpile_cache": True,
     "yaaredis": True,
     "asyncpg": True,
+    "tornado": False,
 }
 
+
+CONTRIB_DEPENDENCIES = {
```
Comment (emmettbutler) [inline ddtrace/_monkey.py:89]: It's weird that this information can't be included in the `tornado` contrib module itself, because that would mean accessing it would require an import. It makes sense the way you have it. An explanatory comment or more detailed variable name could clarify why this is necessary.

File: ddtrace/_monkey.py [hunk 8]
```diff
@@ -84,24 +82,22 @@
...
-# DEV: <contrib name> => <list of module names that trigger a patch>
-_PATCH_ON_IMPORT = {
-    "aiohttp": ("aiohttp",),
-    "aiobotocore": ("aiobotocore",),
-    "celery": ("celery",),
-    "flask": ("flask",),
-    "gevent": ("gevent",),
-    "requests": ("requests",),
-    "botocore": ("botocore",),
+# Module names that need to be patched for a given integration. If the module
```
Comment (emmettbutler) [inline ddtrace/_monkey.py:97]: How was `snowflake.connector` getting patched prior to this change given that this dict didn't exist?

File: ddtrace/_monkey.py [hunk 9]
```diff
@@ -179,11 +180,13 @@ def patch_all(**patch_modules):
...
-        if env_var not in os.environ:
-            continue
+        if env_var in os.environ:
+            modules[module] = formats.asbool(os.environ[env_var])
 
-        override_enabled = formats.asbool(os.environ[env_var])
-        modules[module] = override_enabled
+        # Enable all dependencies for the module
+        if modules[module]:
+            for dep in CONTRIB_DEPENDENCIES.get(module, ()):
```
Comment (emmettbutler) [inline ddtrace/_monkey.py:188]: This is an elegant solution to the cross-contrib dependency issue.

File: ddtrace/_monkey.py [hunk 10]
```diff
@@ -200,7 +203,11 @@ def patch_iast(**patch_modules):
     """
     iast_enabled = formats.asbool(os.environ.get(IAST_ENV, "false"))
     if iast_enabled:
-        patch(raise_errors=False, patch_modules_prefix="ddtrace.appsec.iast.taint_sinks", **patch_modules)
+        # TODO: Devise the correct patching strategy for IAST
+        for module in (m for m, e in patch_modules.items() if e):
+            when_imported("hashlib")(
```
Comment (emmettbutler) [inline ddtrace/_monkey.py:208]: Why `hashlib` here? This would benefit from clarification in the docstring.

File: ddtrace/_monkey.py [hunk 11]
```diff
@@ -212,98 +219,35 @@ def patch(raise_errors=True, patch_modules_prefix=DEFAULT_MODULES_PREFIX, **patc
 
         >>> patch(psycopg=True, elasticsearch=True)
     """
-    modules = [m for (m, should_patch) in patch_modules.items() if should_patch]
-    for module in modules:
-        if module in _PATCH_ON_IMPORT:
-            modules_to_poi = _PATCH_ON_IMPORT[module]
-            for m in modules_to_poi:
-                # If the module has already been imported then patch immediately
-                if m in sys.modules:
```
Comment (emmettbutler) [inline ddtrace/_monkey.py:221]: Checking my understanding: this condition is removed because your removal of cross-contrib imports means that it will never be True. Is that right?

**Review by emmettbutler — COMMENTED** [review]
@P403n1x87 Checking my understanding: you've basically changed things such that ddtracepy does not proactively import any of the modules for which there are integrations in contrib. If that's correct, I'm on board with this change. I left a bunch of clarifying comments and suggestions. Please let me know if anything's unclear and when you'd like another review.

File: ddtrace/contrib/cassandra/session.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/cassandra/session.py:7]: I haven't checked whether it was also failing before. But we could certainly take this out into its own PR.

File: ddtrace/_monkey.py (hunk 10 above)
Comment (P403n1x87) [reply ddtrace/_monkey.py:208]: This is just the quickest way to make the patching happen on import. I've added a comment there to remind myself to also check with ASM if this would work with them, or if we should have a better pattern for this sort of things.

File: ddtrace/_monkey.py (hunk 11 above)
Comment (P403n1x87) [reply ddtrace/_monkey.py:221]: Correct! If the module is already imported `when_imported` will trigger the hook immediately anyway.

File: ddtrace/contrib/cassandra/__init__.py [hunk 12]
```diff
@@ -28,7 +28,7 @@
 
 with require_modules(required_modules) as missing_modules:
     if not missing_modules:
-        from .session import patch
+        from .patch import patch
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/cassandra/__init__.py:31]: Yes. It was giving some import issues which were fixed by importing from the source directly.

File: ddtrace/contrib/httpx/patch.py [hunk 13]
```diff
@@ -154,13 +155,16 @@ def patch():
...
+    version = parse_version(httpx.__version__)
     _w(httpx.Client, "send", _wrapped_sync_send)
 
     pin = Pin()
-    pin.onto(httpx.AsyncClient)
     pin.onto(httpx.Client)
 
+    if version >= (0, 10):
+        _w(httpx.AsyncClient, "send", _wrapped_async_send)
+        pin.onto(httpx.AsyncClient)
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/httpx/patch.py:166]: It looks like we were doing something wrong here that has been surfaced by paching on import. This is being addressed in #4344

File: ddtrace/internal/remoteconfig/client.py [hunk 14]
```diff
@@ -442,10 +443,10 @@ def request(self):
             self._process_response(response)
         except RemoteConfigError as e:
             self._last_error = str(e)
-            log.debug("remote configuration client reported an error", exc_info=True)
-        except ValueError as e:
-            log.debug("Unexpected response data: %s", e)  # noqa: G200
-        except Exception as e:
-            log.debug("Unexpected error: %s", e)  # noqa: G200
+            log.warning("remote configuration client reported an error", exc_info=True)
+        except (compat.httplib.HTTPException, OSError, IOError):
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/internal/remoteconfig/client.py:447]: No. This was here to reduce the log-noise from RCM. I think we can revert this now, or take it out into its own PR if we think there's any value in handling exceptions this way here

File: tests/contrib/django/test_django.py [hunk 15]
```diff
@@ -1605,6 +1605,7 @@ def test_django_use_handler_resource_format_env(client, test_spans):
                 "-c",
                 (
                     "from ddtrace import config, patch_all; patch_all(); "
+                    "import django; "
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply tests/contrib/django/test_django.py:1608]: Correctamundo!

File: tests/integration/test_integration.py [hunk 16]
```diff
@@ -851,25 +851,6 @@ def test_ddtrace_run_startup_logging_injection(ddtrace_run_python_code_in_subpro
     assert b"ValueError: Formatting field not found in record: 'dd.service'" not in err
 
 
-def test_no_module_debug_log(ddtrace_run_python_code_in_subprocess):
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply tests/integration/test_integration.py:854]: With this refactoring we no longer see debug logs as the ones checked for in this test

File: ddtrace/contrib/cassandra/session.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (majorgreys) [reply ddtrace/contrib/cassandra/session.py:7]: Agreed a bug fix for this would be good.

**Review by Yun-Kim — COMMENTED** [review]
I think the cassandra_cluster, dogpile_cache, httpx, and remote config changes can be pulled out into separate PRs that we can review and approve much faster, since they're unrelated and simple on their own. Otherwise good to go forward with this PR 😄

File: ddtrace/contrib/snowflake/patch.py [hunk 17]
```diff
@@ -34,24 +34,34 @@ def _set_post_execute_tags(self, span):
...
 
 def patch():
-    import snowflake.connector
+    try:
+        import snowflake.connector as c
+    except AttributeError:
+        import sys
 
-    if getattr(snowflake.connector, "_datadog_patch", False):
+        c = sys.modules.get("snowflake.connector")
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/snowflake/patch.py:42]: yep

File: ddtrace/contrib/dogpile_cache/patch.py [hunk 18]
```diff
@@ -1,4 +1,9 @@
-import dogpile
+try:
+    import dogpile.cache as dogpile_cache
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/dogpile_cache/patch.py:2]: Same issue. Dedicated PR would be good.

File: tests/integration/test_integration.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Yun-Kim) [reply tests/integration/test_integration.py:854]: @emmettbutler https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/_monkey.py#L259 and https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/_monkey.py#L303 were deleted in this refactor, so this test will now fail in this PR

**Review by emmettbutler — APPROVED** [review]

**Comment by emmettbutler** [comment]
@P403n1x87 could you add a bit to the description about the risks of this change? I'm having some trouble formulating them myself. It's one of the last bits we need for merge.

**Review by emmettbutler — APPROVED** [review]

**Review by Yun-Kim — APPROVED** [review]
